### PR TITLE
Use the time types instead of a u64

### PR DESCRIPTION
**Pull request by dlrobertson**
Use the [Instant] and [Duration] types defined in #110

This isn't nearly done. I simply use `total_millis` where we should probably continue to pass the time type along. I'd like to get feedback and make sure I'm going the right direction before making all of the changes.

[Instant]: https://github.com/m-labs/smoltcp/blob/master/src/time.rs#L29
[Duration]: https://github.com/m-labs/smoltcp/blob/master/src/time.rs#L102

### CODE REVIEW

File: src/time.rs [hunk 1]
```diff
@@ -62,6 +67,13 @@ impl Instant {
     pub fn total_millis(&self) -> i64 {
         self.millis
     }
+
+    /// Return the [Duration] between two [Instant's].
+    ///
+    /// [Duration] struct.Duration.html
+    pub fn duration_since(&self, other: Instant) -> Duration {
```
Comment (whitequark) [inline src/time.rs]: Shouldn't this be `impl Sub<Instant> for Instant`?

File: src/time.rs [hunk 2]
```diff
@@ -85,6 +97,12 @@ impl ops::Add<Duration> for Instant {
     }
 }
 
+impl ops::AddAssign<Duration> for Instant {
```
Comment (whitequark) [inline src/time.rs]: I think these should go in a separate commit.

File: src/time.rs (hunk 1 above)
Comment (dlrobertson) [reply src/time.rs]: 👍

File: src/time.rs (hunk 2 above)
Comment (dlrobertson) [reply src/time.rs]: :+1:

**Comment by dlrobertson** [comment]
@whitequark would you like all the changes to the TCP timestamps etc to be in this PR too? If so, I'm assuming they should be in a separate commit

**Comment by whitequark** [comment]
@dlrobertson Sure, separate commit seems fine.

File: examples/benchmark.rs [hunk 3]
```diff
@@ -31,7 +31,7 @@ fn client(kind: Client) {
     let mut stream = TcpStream::connect(("192.168.69.1", port)).unwrap();
     let mut buffer = vec![0; 1_000_000];
 
-    let start = Instant::now();
+    let start = Instant::now().expect("Could not get the start time");
```
Comment (whitequark) [inline examples/benchmark.rs]: Hang on. Why would this fail?

File: examples/utils.rs [hunk 4]
```diff
@@ -18,14 +18,15 @@ use smoltcp::phy::{Device, EthernetTracer, FaultInjector};
...
 
 #[cfg(feature = "log")]
 pub fn setup_logging_with_clock<F>(filter: &str, since_startup: F)
-        where F: Fn() -> u64 + Send + Sync + 'static {
+        where F: Fn() -> Instant + Send + Sync + 'static {
     LogBuilder::new()
         .format(move |record: &LogRecord| {
             let elapsed = since_startup();
-            let timestamp = format!("[{:6}.{:03}s]", elapsed / 1000, elapsed % 1000);
+            let timestamp = format!("{}", elapsed);
```
Comment (whitequark) [inline examples/utils.rs]: I think you forgot `[]`.

File: examples/benchmark.rs (hunk 3 above)
Comment (dlrobertson) [reply examples/benchmark.rs]: It makes use of [SystemTime::duration_since] which returns a result type. Now that I'm looking at it, would it make more sense if `Instant::from_system_time` used [SystemTIme::elapsed] instead of [SystemTime::duration_since] the [UNIX_EPOCH]? Or would it make sense to have a `Instant::elapsed()`?

[SystemTime::duration_since]: https://doc.rust-lang.org/std/time/struct.SystemTime.html#method.duration_since
[SystemTime::elapsed]: https://doc.rust-lang.org/std/time/struct.SystemTime.html#method.elapsed
[UNIX_EPOCH]: https://doc.rust-lang.org/std/time/constant.UNIX_EPOCH.html

Comment (whitequark) [reply examples/benchmark.rs]: It should just use `duration_since(UNIX_EPOCH).unwrap()`. I am not interested in handling cases where the time is less than UNIX epoch. The only such case is badly misconfigured (i.e. non-UTC timezone) systems without an RTC battery or ntpd.

Comment (whitequark) [reply examples/benchmark.rs]: Actually, nevermind, not even that can happen because UNIX_EPOCH includes the timezone. So just unwrap it.

Comment (dlrobertson) [reply examples/benchmark.rs]: :+1:

**Comment by dlrobertson** [comment]
Updated all the things!

Added commits that update the socket modules to use `time` etc.

File: src/time.rs [hunk 5]
```diff
@@ -23,7 +23,7 @@ use core::{ops, fmt};
 ///   point.
 #[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
 pub struct Instant {
-    millis: i64,
+    pub millis: i64,
```
Comment (dlrobertson) [inline src/time.rs]: Updated to `pub` so that constants can be used.

File: examples/benchmark.rs [hunk 6]
```diff
@@ -52,11 +52,9 @@ fn client(kind: Client) {
 
     let end = Instant::now();
 
-    let elapsed = end - start;
-    let elapsed = elapsed.as_secs() as f64
-                + elapsed.subsec_nanos() as f64 * 1e-9;
+    let elapsed = (end - start).total_millis() as f64 / 1000.0;
 
-    println!("throughput: {:.3} Gbps", AMOUNT as f64 / elapsed / 0.125e9);
+    println!("throughput: [{:.3}] Gbps", AMOUNT as f64 / elapsed / 0.125e9);
```
Comment (whitequark) [inline examples/benchmark.rs]: huh?

File: examples/loopback.rs [hunk 7]
```diff
@@ -57,12 +60,12 @@ mod mock {
...
-        pub fn advance(&self, millis: u64) {
-            self.0.fetch_add(millis as usize, Ordering::SeqCst);
+        pub fn advance(&self, duration: Duration) {
+            self.0.fetch_add(duration.total_millis() as usize, Ordering::SeqCst);
         }
 
-        pub fn elapsed(&self) -> u64 {
-            self.0.load(Ordering::SeqCst) as u64
+        pub fn elapsed(&self) -> Instant {
+			Instant::from_millis(self.0.load(Ordering::SeqCst) as i64)
```
Comment (whitequark) [inline examples/loopback.rs]: something weird happened with indentation

File: examples/loopback.rs [hunk 8]
```diff
@@ -164,13 +167,15 @@ fn main() {
         }
 
         match iface.poll_delay(&socket_set, clock.elapsed()) {
-            Some(0) =>
```
Comment (whitequark) [inline examples/loopback.rs]: now that `millis` is `pub`, I think you can match directly

File: examples/benchmark.rs (hunk 6 above)
Comment (dlrobertson) [reply examples/benchmark.rs]: updated the wrong line. fixed.

File: examples/loopback.rs (hunk 7 above)
Comment (dlrobertson) [reply examples/loopback.rs]: That is strange... somehow I inserted a tab here. Not sure how that happened.

File: examples/loopback.rs (hunk 8 above)
Comment (dlrobertson) [reply examples/loopback.rs]: Good point.

**Comment by dlrobertson** [comment]
A few things before this should be merged.

 - `From<::std::time::Instant>` isn't implemented. Adding this now
 - `from_system_time` should be `From<::std::time::SystemTime>` now that it isn't a result type.
 - Question: Should `Instant::now()` be a "wrapper" for `::std::time::Instant::now()` or what is currently done with the `UNIX_EPOCH`?

**Comment by dlrobertson** [comment]
Updated and I didn't change `Instant::now`. I think sticking with the current implementation makes more sense.

File: src/phy/fault_injector.rs [hunk 9]
```diff
@@ -17,21 +18,35 @@ fn xorshift32(state: &mut u32) -> u32 {
...
     drop_pct:    u8,
     reorder_pct: u8,
     max_size:    usize,
     max_tx_rate: u64,
     max_rx_rate: u64,
-    interval:    u64,
+    interval:    Duration,
+}
+
+impl Default for Config {
```
Comment (whitequark) [inline src/phy/fault_injector.rs]: Why not `impl Default for Duration`? `std` does this.

File: src/time.rs [hunk 10]
```diff
@@ -32,17 +32,15 @@ impl Instant {
...
-    #[cfg(feature = "std")]
-    pub fn now() -> Result<Instant, ::std::time::SystemTimeError> {
-        Self::from_system_time(::std::time::SystemTime::now())
+    /// Create a new `Instant` from a number of seconds.
+    pub fn from_secs(secs: i64) -> Instant {
+        Instant { millis: secs * 1000 }
     }
 
-    /// Create a new `Instant` from a `SystemTime`.
+    /// Create a new `Instant` from the current `SystemTime`.
```
Comment (whitequark) [inline src/time.rs]: Nit: it's not clear from the signature and doc comment that this means `std::time::SystemTime`.

**Comment by whitequark** [comment]
@m-labs-homu delegate+
@dlrobertson r=me once the comments above are fixed

**Comment by dlrobertson** [comment]
@m-labs-homu r=whitequark

**Comment by m-labs-homu** [comment]
:pushpin: Commit cbedd74 has been approved by `whitequark`

**Comment by m-labs-homu** [comment]
:hourglass: Testing commit cbedd743db9bef88e4e0dd8e15bf042fa38c0ae4 with merge c418b60b5db93753999ae51d33ec4dc1d1631c69...

**Comment by m-labs-homu** [comment]
:sunny: Test successful - [status-travis](https://travis-ci.org/m-labs/smoltcp/builds/342175985?utm_source=github_status&utm_medium=notification)
Approved by: whitequark
Pushing c418b60b5db93753999ae51d33ec4dc1d1631c69 to master...

**Comment by whitequark** [comment]
Thanks! Great work.
